src/util/logger.rs

```diff
@@ -0,0 +1,389 @@
+				.position(|&name| str::eq_ignore_ascii_case(name, level))
+				.into_iter()
+				.filter(|&idx| idx != 0)
+				.map(|idx| Level::from_usize(idx).unwrap())
+				.next(),
+			ParseLevelError(()),
+		)
+	}
+}
+
+impl fmt::Display for Level {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		fmt.pad(LOG_LEVEL_NAMES[*self as usize])
+	}
+}
+
+#[allow(dead_code)]
+impl Level {
+	fn from_usize(u: usize) -> Option<Level> {
+		match u {
+			1 => Some(Level::Error),
+			2 => Some(Level::Warn),
+			3 => Some(Level::Info),
+			4 => Some(Level::Debug),
+			5 => Some(Level::Trace),
+			_ => None,
+		}
+	}
+
+	/// Returns the most verbose logging level.
+	#[inline]
+	pub fn max() -> Level {
+		Level::Trace
+	}
+}
+
+/// A Record, unit of logging output with Metadata to enable filtering
+/// Module_path, file, line to inform on log's source
+#[derive(Clone,Debug)]
+pub struct Record<'a> {
+	metadata: Metadata<'a>,
+	args: fmt::Arguments<'a>,
+	module_path: Option<&'a str>,
+	file: Option<&'a str>,
+	line: Option<u32>,
+}
+
+#[allow(dead_code)]
+impl<'a> Record<'a> {
+	/// Returns a new Record.
+	#[inline]
+	pub fn new(metadata: Metadata<'a>, args: fmt::Arguments<'a>, module_path: Option<&'a str>, file: Option<&'a str>, line: Option<u32>) -> Record<'a> {
+		Record {
+			metadata: metadata,
+			args: args,
+			module_path: module_path,
+			file: file,
+			line: line
+		}
+	}
+
+	/// The message body.
+	#[inline]
+	pub fn args(&self) -> &fmt::Arguments<'a> {
+		&self.args
+	}
+
+	/// Metadata about the log directive.
+	#[inline]
+	pub fn metadata(&self) -> &Metadata<'a> {
+		&self.metadata
+	}
+
+	/// The verbosity level of the message.
+	#[inline]
+	pub fn level(&self) -> Level {
+		self.metadata.level()
+	}
+
+	/// The name of the target of the directive.
+	#[inline]
+	pub fn target(&self) -> &'a str {
+		self.metadata.target()
+	}
+
+	/// The module path of the message.
+	#[inline]
+	pub fn module_path(&self) -> Option<&'a str> {
+		self.module_path
+	}
+
+	/// The source file containing the message.
+	#[inline]
+	pub fn file(&self) -> Option<&'a str> {
+		self.file
+	}
+
+	/// The line containing the message.
+	#[inline]
+	pub fn line(&self) -> Option<u32> {
+		self.line
+	}
+}
+
+/// Metadata about a log message
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct Metadata<'a> {
+	level: Level,
+	target: &'a str,
+}
+
+#[allow(dead_code)]
+impl<'a> Metadata<'a> {
+	/// Returns a new Metadata.
+	#[inline]
+	pub fn new(level: Level, target: &'a str) -> Metadata<'a> {
+		Metadata {
+			level,
+			target,
+		}
+	}
+
+	/// The verbosity level of the message.
+	#[inline]
+	pub fn level(&self) -> Level {
+		self.level
+	}
+
+	/// The name of the target of the directive.
+	#[inline]
+	pub fn target(&self) -> &'a str {
+		self.target
+	}
+}
+
+/// A trait encapsulating the operations required of a logger
+pub trait Logger: Sync + Send {
+	/// Enable the logger with a maximum log level
+	fn enable(&mut self, level: Level);
```

Comment:
Why does this (and flush, and level) need to be present? I think that's all client-side stuff that a library doesn't really need to concern itself with?

Reply:
I'm still unsure why we'd need a third way to limit the log level (which this appears to be?).